logging/jsonlog/jsonlog.go

```diff
@@ -0,0 +1,327 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package jsonlog provides a Logger that logs structured JSON to Stderr by
+// default. When used on the various Cloud Compute environments (Cloud Run,
+// Cloud Functions, GKE, etc.) these JSON messages will be parsed by the Cloud
+// Logging agent and transformed into a message format that mirrors that of the
+// Cloud Logging API.
+package jsonlog
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/logging"
+	"cloud.google.com/go/logging/internal"
+	logtypepb "google.golang.org/genproto/googleapis/logging/type"
+	logpb "google.golang.org/genproto/googleapis/logging/v2"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+// NewLogger creates a Logger that logs structured JSON to Stderr. The value of
+// parent must be in the format of:
+//    projects/PROJECT_ID
+//    folders/FOLDER_ID
+//    billingAccounts/ACCOUNT_ID
+//    organizations/ORG_ID
+func NewLogger(parent string, opts ...LoggerOption) (*Logger, error) {
+	return NewLoggerFromRequest(parent, nil, opts...)
+}
+
+// NewLoggerFromRequest creates a Logger similar to NewLogger but with the
+// additional context that the current request can provide.
+// The value of parent must be in the format of:
+//    projects/PROJECT_ID
+//    folders/FOLDER_ID
+//    billingAccounts/ACCOUNT_ID
+//    organizations/ORG_ID
+func NewLoggerFromRequest(parent string, r *http.Request, opts ...LoggerOption) (*Logger, error) {
+	err := validateParent(parent)
+	if err != nil {
+		return nil, err
+	}
+	var req *logtypepb.HttpRequest
+	if r != nil {
+		u := *r.URL
+		req = &logtypepb.HttpRequest{
+			RequestMethod: r.Method,
+			RequestUrl:    internal.FixUTF8(u.String()),
+			UserAgent:     r.UserAgent(),
+			Referer:       r.Referer(),
+			Protocol:      r.Proto,
+		}
+		if r.Response != nil {
+			req.Status = int32(r.Response.StatusCode)
+		}
+	}
+	l := &Logger{
+		w:   os.Stderr,
+		req: req,
+	}
+
+	var traceHeader string
+	if r != nil && r.Header != nil {
+		traceHeader = r.Header.Get("X-Cloud-Trace-Context")
+	}
+	if traceHeader != "" {
+		traceID, spanID, traceSampled := internal.DeconstructXCloudTraceContext(traceHeader)
+		l.traceID = fmt.Sprintf("%s/traces/%s", parent, traceID)
+		l.spanID = spanID
+		l.sampled = traceSampled
+	}
+	for _, opt := range opts {
+		opt.set(l)
+	}
+	return l, nil
+}
+
+// Logger is used for logging JSON entries.
+type Logger struct {
+	w       io.Writer
+	now     func() time.Time
+	errhook func(error)
+
+	// read-only fields
+	labels  map[string]string
+	req     *logtypepb.HttpRequest
+	traceID string
+	sampled bool
+	spanID  string
+}
+
+// Debugf is a convenience method for writing an Entry with a Debug Severity
+// and the provided formatted message.
+func (l *Logger) Debugf(format string, a ...interface{}) {
+	e := entry{
+		Message:  fmt.Sprintf(format, a...),
+		Severity: "DEBUG",
```

Comment:
Does it make sense for the severity strs to be declared as constants? Not a go expert, asking more from readability  standpoint.